fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java

```diff
@@ -74,228 +99,255 @@ public void setup() throws GeneralSecurityException, IOException {
+    Set<QuestionResult> results = Collections.singleton(new QuestionResult(OPTION_1, 1));
+
+    handleElectionSetup(OPEN_BALLOT_ELECTION);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+    handleCastVote(new ElectionVote(QUESTION.getId(), 0, false, null, ELECTION_ID), SENDER_KEY);
+    handleElectionEnd();
+    handleElectionResults(results, OPEN_BALLOT_ELECTION.getChannel());
+
+    Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
+    assertEquals(EventState.RESULTS_READY, election.getState());
+    assertEquals(results, election.getResultsForQuestionId(QUESTION.getId()));
   }
 
   @Test
-  public void testHandleCastVote()
+  public void testHandleElectionOpen()
       throws DataHandlingException, UnknownLaoException, UnknownRollCallException,
-          NoRollCallException {
-    // Here we test if the handling can both an Open Ballot cast vote and
-    // an Secret Ballot vote
-
-    // First test the Open Ballot version
-    // Set up a open ballot election
-    ElectionVote electionVote1 = new ElectionVote("1", 1, false, null, election.getId());
-    List<ElectionVote> electionVotes = Collections.singletonList(electionVote1);
-    CastVote<ElectionVote> electionVote =
-        new CastVote<>(electionVotes, election.getId(), CREATE_LAO.getId());
+          UnknownElectionException, NoRollCallException {
+    handleElectionSetup(OPEN_BALLOT_ELECTION);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+
+    Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
+    assertEquals(OPENED_AT, election.getStartTimestamp());
+    assertEquals(EventState.OPENED, election.getState());
+  }
 
-    MessageGeneral message1 = new MessageGeneral(SENDER_KEY, electionVote, gson);
-    // Test the whole process
-    messageHandler.handleMessage(messageSender, LAO_CHANNEL.subChannel(election.getId()), message1);
-    List<String> listOfVoteIds = new ArrayList<>();
-    // Since messageMap is a TreeMap, votes will already be sorted in the alphabetical order of
-    // messageIds
-    listOfVoteIds.add(electionVote1.getId());
-    String expectedHash = Hash.hash(listOfVoteIds.toArray(new String[0]));
-    assertEquals(expectedHash, election.computerRegisteredVotes());
+  @Test
+  public void testHandleElectionEnd()
+      throws DataHandlingException, UnknownLaoException, UnknownRollCallException,
+          UnknownElectionException, NoRollCallException {
+    handleElectionSetup(OPEN_BALLOT_ELECTION);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+    handleElectionEnd();
+
+    Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
+    assertEquals(EventState.CLOSED, election.getState());
+  }
 
-    // Now test with a SECRET BALLOT election
```

Comment:
The downside of all this refactoring is that we part ways with some great comments 😆